### ISSUE TITLE
"Blame previous revision" menu item always disabled

### ISSUE THREAD

**Issue by shirayuki** [gitextensions/gitextensions#10946]
### Environment

- Git Extensions 4.0.2.16100
- Build 25100ec1f7c8da8f798613da74826038af81a50e
- Git 2.38.1.windows.1
- Microsoft Windows NT 10.0.22621.0
- .NET 6.0.13
- DPI 96dpi (no scaling)
- Microsoft.WindowsDesktop.App Versions

```
    Microsoft.WindowsDesktop.App 5.0.5 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
    Microsoft.WindowsDesktop.App 6.0.13 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

### Issue description

"Blame previous revision" menu item is always disabled in the Blame pane

### Steps to reproduce

1. Right-click on a file in the repository in Windows Explorer.
2. Select "File history" from the context menu.
3. In the File History window, right-click on a revision in the Blame pane.
4. A context menu will appear. The menu item "Blame previous revision" should be present.
5. Notice that this menu item is always disabled, even when there are previous revisions to the file.

**Comment by mstv** [comment]
The menu item is disabled if that revision does not exist in the revision grid.
You could try to disable some filters (e.g. simplifications, merge commits) or also the number of loaded commits.

**Comment by shirayuki** [comment]
> The menu item is disabled if that revision does not exist in the revision grid. You could try to disable some filters (e.g. simplifications, merge commits) or also the number of loaded commits.

Thank you for your response.
I understand that the issue is related to the revision grid and the "Blame previous revision" menu item being disabled when the revision is not present in the grid.
Based on your explanation, it seems that the problem occurs because the file history view is only showing revisions for the selected file, and the target revision for "Blame previous revision" may not be included in that view.
Thank you for clarifying this for me.

**Comment by gerhardol** [comment]
You can blame a file in the main Browse window without filtering.
All versions of a file may not be shown in the revision grid so "previous revision" of the file would not be displayed. But it is maybe OK to change this if the caption is changed to "Blame previous revision in grid".

**Comment by fschmied** [comment]
+ Even if you know the reason, resetting the path filter at this point to enable "Blame previous revision" will lose the file context when "Show file history in the main window" is enabled: ![Image](https://github.com/user-attachments/assets/3d5673ab-6593-4db9-a277-ceaf5c9f1acb) and you need to manually search for the file again.

(This actually works better when switching back to the mode with a "separate" File History, which doesn't lose the file context. I find that mode easier to work with anyway; to me, it's much less confusing.)

**Comment by fschmied** [comment]
> (This actually works better when switching back to the mode with a "separate" File History, which doesn't lose the file context. I find that mode easier to work with anyway; to me, it's much less confusing.)

OTOH, I just noticed that the old "File History" version gets really confused when repeating "Blame previous" over moves...
![Image](https://github.com/user-attachments/assets/77e6c978-4c9e-41e7-897b-2bb3f1ba1823)

It will no longer be able to find the file:
![Image](https://github.com/user-attachments/assets/78678811-6145-4d23-8bcd-2eab4cf700e8)
![Image](https://github.com/user-attachments/assets/3113cff6-e2df-4642-b717-f074c6a1aaa0)

So, not really a good workaround, either.

**Comment by mstv** [comment]
I agree that using the filter for the old File History does not work (well) for blame.

The `View` tab tells what went wrong:
`'Cannot get treeId from Git for "<file.x>" for commit <hash>.'`

Blame works best in the `"File tree"` tab of the main window (hit `t`, `b` and if filtered, `Shift+Ctrl+I` in order to disable filters).

---

> (Note that the history appears broken at first as the panel at the bottom doesn't show anything... But that's not in scope of this issue here.)

(It is not broken. But no revision is selected, which can be improved.)

**Comment by gerhardol** [comment]
> (Note that the history appears broken at first as the panel at the bottom doesn't show anything... But that's not in scope of this issue here.) 

GE tries to select new HEAD if HEAD was selected before or the previous selected commit, if that is visible.
It could be confusing to select something "random" too. I guess the top commit if HEAD was selected would not be too annoying.

> Every change has "Blame previous revision" disabled... 

The previous revision is normally not seen, so that is expected.
Previous commit in grid is though, but it was not trivial if I remember.

> It will no longer be able to find the file:

File tree has no filter, so it is hard to follow. Git has no simple way to handle rename, it is a lot of implementation to track changes.
You can go from Diff tab to file tree to display files, the diff files may be fewer an can be filtered.
I guess it is possible to use the list of all renamed names to have a priority list with files to "follow" when selecting a new commit.
The order is now the filtered path only, then last explicitly selected, then last displayed. For Diff, maybe for file tree as well.

**Comment by gerhardol** [comment]
> `Blame previous revision` should always jump to the previous commit loaded in the Revision Grid.

I suggest a new menu item, in addition to current

**Comment by mstv** [comment]
> > `Blame previous revision` should always jump to the previous commit loaded in the Revision Grid.
> 
> I suggest a new menu item, in addition to current

The blame view would look the same if viewing the history of that file, (which is the major use-case).
(Using _blame_ with stronger / other filters has always been confusing / disappointing to me.)
What would be the benefit of a yet another menu item from user perspective?

The implementation is easy: Select the revision used for "Blame this revision" and then perform what `Ctrl+P` does.

If (we could not restore the previous selection or its predecessor and) the HEAD revision has been filtered out, too, we should always select the first non-artificial commit. (This will also avoid this nasty focus rectangle.)

Have I missed something?

**Comment by gerhardol** [comment]
> The implementation is easy: Select the revision used for "Blame this revision" and then perform what Ctrl+P does.

I just recall something with there were something with it when I very briefly looked at that. But it was a long time ago.

> If (we could not restore the previous selection or its predecessor and) the HEAD revision has been filtered out, too, we should always select the first non-artificial commit. (This will also avoid this nasty focus rectangle.)

If you have selected an old commit, you will jump to the beginning. If there is no commit in the branch you had checked out, you may get a branch from a different branch. Sure you will see something, but maybe something irrelevant. Since context is changed, you could be confused.
Not necessarily better, but I will not vote against the topmost commit.
